flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -349,11 +346,13 @@ public void setNumBuffers(int numBuffers) throws IOException {
 
 			returnExcessMemorySegments();
 
-			// If there is a registered owner and we have still requested more buffers than our
-			// size, trigger a recycle via the owner.
-			if (owner != null && numberOfRequestedMemorySegments > currentPoolSize) {
-				owner.releaseMemory(numberOfRequestedMemorySegments - currentPoolSize);
-			}
+			numReleased = numberOfRequestedMemorySegments - currentPoolSize;
+		}
+
+		// If there is a registered owner and we have still requested more buffers than our
+		// size, trigger a recycle via the owner.
+		if (owner != null && numReleased > 0) {
+			owner.releaseMemory(numReleased);
```

Comment:
Are you sure that doing this outside of the `synchronized` block will not cause some race conditions? After this change from the perspective of different threads interacting with this class, memory will not be release during `setNumBuffers` but at some random point time in the future.

Reply:
Yes, I also have not thought of any potential problems currently.